Practical Example/Neil Hiddink/RWContent/RWContent/Controllers/ContentViewController.swift

```diff
@@ -0,0 +1,205 @@
+        container.frame = view.frame
+        container.center = view.center
+        container.backgroundColor = UIColor(white: 0.3, alpha: 0.3)
+        container.layer.cornerRadius = 10
+        
+        let loadingView: UIView = UIView()
+        loadingView.frame = CGRect(x: 0, y: 0, width: 80, height: 80)
+        loadingView.center = view.center
+        loadingView.backgroundColor = UIColor(white: 0.3, alpha: 0.0)
+        loadingView.clipsToBounds = true
+
+        activityIndicator.style = .large
+        activityIndicator.color = .systemGray6
+        activityIndicator.frame = loadingView.frame
+        activityIndicator.center = CGPoint(x: loadingView.frame.width / 2, y: loadingView.frame.height / 2)
+
+        loadingView.addSubview(activityIndicator)
+        container.addSubview(loadingView)
+        view.addSubview(container)
+        activityIndicator.startAnimating()
+    }
+}
+
+extension ContentViewController: UITableViewDelegate, UITableViewDataSource {
+    
+    // MARK: - Delegate
+    
+    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        
+        let baseURL = "https://www.raywenderlich.com"
+        
+        guard let path = (content[indexPath.row].links["self"] as? String)?.dropFirst(41) else { fatalError() }
+        guard let url = URL(string: baseURL + path) else { fatalError() }
+            
+        let configuration = SFSafariViewController.Configuration()
+        configuration.entersReaderIfAvailable = false
+            
+        // https://www.hackingwithswift.com/read/32/3/how-to-use-sfsafariviewcontroller-to-browse-a-web-page
+        let safariVC = SFSafariViewController(url: url, configuration: configuration)
+        safariVC.modalPresentationStyle = .overFullScreen
+        safariVC.preferredControlTintColor = .gray
+        present(safariVC, animated: true)
+    }
+    
+    // MARK: - Data Source
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 180.0
+    }
+    
+    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if let cell = tableView.dequeueReusableCell(withIdentifier: "RWContentTableViewCell") as? RWContentTableViewCell {
+            let contentItem = content[indexPath.row]
+            cell.titleLabel.text = contentItem.attributes.name
+            cell.descriptionLabel.text = contentItem.attributes.description_plain_text
+            cell.contentTypeLabel.text = "\(contentItem.attributes.content_type.uppercased()) – " +
+                                    "\(contentItem.attributes.difficulty?.uppercased() ?? "ANNOUNCEMENTS")"
+                                    
+            if let imageUrl = URL(string: contentItem.attributes.card_artwork_url) {
+                let data: Data?
+                do {
+                    data = try Data(contentsOf: imageUrl)
```

Comment:
`Data(contentsOf:)`  is a pretty bad way to fetch images, for several reasons:

 - It works on the main thread, which is why your table view is so choppy. Whenever a new cell is about to appear, the main thread stops to download an image from the internet.
 - There's no caching, meaning you have to re-download the same image over and over while scrolling.
 - There's no way to cancel the request if an image goes off-screen.
 - There's no way to usefully handle errors.

Instead of this, I would suggest either making your own image downloader with URLSession or using libraries like [Kingfisher](https://github.com/onevcat/Kingfisher) or [AlamofireImage](https://github.com/Alamofire/AlamofireImage).

Reply:
Great points, I'll have to remember these things in future projects! I did feel like the table view was not going to scale well with more data. Also, I found this library the other day, maybe it's equivalent to the ones you mentioned? https://github.com/hyperoslo/Imaginary

Reply:
Looking at the link it seems like yes, that library does the same things as the ones I mentioned.